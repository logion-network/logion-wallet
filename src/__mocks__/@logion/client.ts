import { CheckResultType } from "@logion/client/dist/CollectionItem.js";
import { DefaultSignAndSendStrategy } from "@logion/client/dist/Signer.js";

import {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    LocRequestState,
    LocsState,
    HashOrContent,
} from '../LogionClientMock';

export {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    CheckResultType,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    DefaultSignAndSendStrategy,
    LocRequestState,
    LocsState,
    HashOrContent,
}

export { MimeType } from "@logion/client/dist/Mime.js";
