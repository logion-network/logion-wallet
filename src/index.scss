@import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@400;600;700;800&display=swap');
@import url('https://fonts.googleapis.com/css2?family=Lexend+Deca&display=swap');

$font-family-sans-serif: 'Montserrat', sans-serif;
$font-family-monospace: "Source Code Pro", monospace !default;
$grid-breakpoints: (
  xs: 0,
  sm: 576px,
  md: 768px,
  lg: 992px,
  xl: 1200px,
  xxl: 1400px,
  xxxl: 1600px
);
$container-max-widths: (
  sm: 540px,
  md: 720px,
  lg: 960px,
  xl: 1140px,
  xxl: 1320px,
  xxxl: 1400px
);


@import "~bootstrap/scss/bootstrap";

body, html {
    height: 100%;
    width: 100%;
}

html {
    font-size: 16px;
}

#root {
    height: 100%;
    width: 100%;
}

@media (max-width: 1350px) {
    html {
        font-size: 14px;
    }
}

.spinner-border.text-polkadot {
    color: #e6007a;
}
