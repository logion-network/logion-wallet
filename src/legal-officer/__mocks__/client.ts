import {
    closeLoc,
    deleteLink,
    voidLoc,
    getVerifiedThirdPartySelections,
} from "./ClientMock";

export {
    closeLoc,
    deleteLink,
    voidLoc,
    getVerifiedThirdPartySelections,
}
