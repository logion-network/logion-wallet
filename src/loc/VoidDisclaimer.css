.VoidDisclaimer.loc-is-void {
    border-top-left-radius: 0;
    border-top-right-radius: 0;
    margin-top: -2px;
}
