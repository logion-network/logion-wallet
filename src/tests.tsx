import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };

export async function clickByName(name: string | RegExp | ((accessibleName: string, element: Element) => boolean)) {
    let button: HTMLElement;
    await waitFor(() => button = screen.getByRole("button", { name }));
    await userEvent.click(button!);
}

export async function typeByLabel(label: string, value: string) {
    let input: HTMLElement;
    await waitFor(() => input = screen.getByLabelText(label));
    await userEvent.type(input!, value);
}
