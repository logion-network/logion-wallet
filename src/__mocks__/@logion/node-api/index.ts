import { ApiPromise } from "src/__mocks__/PolkadotApiMock";
export * from "@logion/node-api/dist/UUID";

export function buildApi() {
    return new ApiPromise();
}

export * from "./dist/LogionLoc";
