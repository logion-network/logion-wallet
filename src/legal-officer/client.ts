import {
    EditableRequest,
    Signer,
    SignCallback,
    LocRequestState,
    OpenLoc,
    ClosedLoc,
    ClosedCollectionLoc,
    VerifiedThirdParty,
    VoidedLoc,
    VoidedCollectionLoc,
    LogionClient,
    LegalOfficer,
} from "@logion/client";
import {
    UUID,
    addLink as nodeApiAddLink,
    addFile as nodeApiAddFile,
    addMetadata as nodeApiAddMetadata,
    MetadataItem,
    closeLoc as nodeApiCloseLoc,
    voidLoc as nodeApiVoidLoc,
    VoidInfo,
    asString,
} from "@logion/node-api";
import { AnyJson } from "@polkadot/types-codec/types/helpers.js";
import { fetchAllLocsParams } from "../loc/LegalOfficerLocContext";

export async function addLink(params: {
    locState: EditableRequest,
    target: UUID,
    nature: string,
}): Promise<EditableRequest> {
    const { locState, target, nature } = params;
    const axios = buildAxios(locState);
    await axios.post(`/api/loc-request/${ locState.data().id.toString() }/links`, { target: target.toString(), nature })
    return await locState.refresh() as EditableRequest;
}

export async function deleteLink(params: {
    locState: EditableRequest,
    target: UUID,
}): Promise<EditableRequest> {
    const { locState, target } = params;
    const axios = buildAxios(locState);
    await axios.delete(`/api/loc-request/${ locState.data().id.toString() }/links/${target}`);
    return await locState.refresh() as EditableRequest;
}

function buildAxios(locState: LocRequestState) {
    const client = locState.locsState().client;
    return client.buildAxios(getLegalOfficer(client, locState.data().ownerAddress));
}

function getLegalOfficer(client: LogionClient, address: string): LegalOfficer {
    return client.legalOfficers.find(legalOfficer => address === legalOfficer.address)!;
}

export async function publishLink(params: {
    locState: EditableRequest,
    target: UUID,
    nature: string,
    signer: Signer,
    callback: SignCallback,
}): Promise<EditableRequest> {
    const { locState, target, nature, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locId = locState.data().id;
    const submittable = nodeApiAddLink({
        locId,
        api,
        target,
        nature,
    });
    await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });

    const axios = buildAxios(locState);
    await axios.put(`/api/loc-request/${ locId.toString() }/links/${ target.toString() }/confirm`);

    return await locState.refresh() as EditableRequest;
}

export async function publishFile(params: {
    locState: EditableRequest,
    hash: string,
    nature: string,
    submitter: string,
    signer: Signer,
    callback: SignCallback,
}): Promise<EditableRequest> {
    const { locState, hash, nature, submitter, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locId = locState.data().id;
    const submittable = nodeApiAddFile({
        locId,
        api,
        hash,
        nature,
        submitter,
    });
    await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });

    const axios = buildAxios(locState);
    await axios.put(`/api/loc-request/${ locId.toString() }/files/${ hash }/confirm`);

    return await locState.refresh() as EditableRequest;
}

export async function publishMetadata(params: {
    locState: EditableRequest,
    item: MetadataItem,
    signer: Signer,
    callback: SignCallback,
}): Promise<EditableRequest> {
    const { locState, item, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locId = locState.data().id;
    const submittable = nodeApiAddMetadata({
        locId,
        api,
        item,
    });
    await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });

    const axios = buildAxios(locState);
    await axios.put(`/api/loc-request/${ locId.toString() }/metadata/${ encodeURIComponent(item.name) }/confirm`);

    return await locState.refresh() as EditableRequest;
}

export async function closeLoc(params: {
    locState: OpenLoc,
    signer: Signer,
    callback: SignCallback,
}): Promise<ClosedLoc | ClosedCollectionLoc> {
    const { locState, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locData = locState.data();
    const locId = locData.id;
    const seal = locData.seal;
    const submittable = nodeApiCloseLoc({
        locId,
        api,
        seal,
    });
    const signerId = locState.data().ownerAddress;
    await signer.signAndSend({
        signerId,
        submittable,
        callback
    });

    let refreshedLocState;
    if(locState.discarded) {
        refreshedLocState = (await client.locsState(fetchAllLocsParams(getLegalOfficer(client, signerId)))).findById(locId);
    } else {
        refreshedLocState = locState;
    }

    const axios = buildAxios(refreshedLocState);
    await axios.post(`/api/loc-request/${ locId.toString() }/close`);

    return await refreshedLocState.refresh() as ClosedLoc | ClosedCollectionLoc;
}

export interface FullVoidInfo extends VoidInfo {
    reason: string;
}

export async function voidLoc(params: {
    locState: LocRequestState,
    voidInfo: FullVoidInfo,
    signer: Signer,
    callback: SignCallback,
}): Promise<VoidedLoc | VoidedCollectionLoc> {
    const { locState, voidInfo, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locData = locState.data();
    const locId = locData.id;
    const submittable = nodeApiVoidLoc({
        locId,
        api,
        voidInfo,
    });
    await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });

    const axios = buildAxios(locState);
    const reason = voidInfo.reason;
    await axios.post(`/api/loc-request/${ locId.toString() }/void`, {
        reason
    });

    return await locState.refresh() as VoidedLoc | VoidedCollectionLoc;
}

export async function setVerifiedThirdParty(params: {
        locState: ClosedLoc,
        isVerifiedThirdParty: boolean;
    }
): Promise<ClosedLoc> {
    const { locState, isVerifiedThirdParty } = params;
    const axios = buildAxios(locState);
    await axios.put(`/api/loc-request/${locState.data().id.toString()}/verified-third-party`, {
        isVerifiedThirdParty
    });
    return await locState.refresh() as ClosedLoc;
}

export async function getVerifiedThirdPartySelections(params: { locState: OpenLoc } ): Promise<VerifiedThirdParty[]> {
    const { locState } = params
    const axios = buildAxios(locState);
    const response = await axios.get("/api/verified-third-parties");
    const allVerifiedThirdParties: VerifiedThirdParty[] = response.data.verifiedThirdParties;

    const selectedParties = locState.data().selectedParties;

    return allVerifiedThirdParties
        .filter(vtp => vtp.address !== locState.data().requesterAddress)
        .map(vtp => selectedParties.find(selectedParty => selectedParty.identityLocId === vtp.identityLocId && selectedParty.selected) ?
            { ...vtp, selected: true } :
            { ...vtp, selected: false }
        )
        .sort((vtp1, vtp2) => vtp1.lastName.localeCompare(vtp2.lastName));
}

export interface SelectPartiesParams {
    locState: OpenLoc;
    partyId: string;
}

export async function selectParties(params: SelectPartiesParams): Promise<void> {
    const { locState, partyId } = params;
    const axios = buildAxios(locState);
    await axios.post(`/api/loc-request/${locState.locId.toString()}/selected-parties`, {
        identityLocId: partyId
    })
}

export async function unselectParties(params: SelectPartiesParams): Promise<void> {
    const { locState, partyId } = params;
    const axios = buildAxios(locState);
    await axios.delete(`/api/loc-request/${ locState.locId.toString() }/selected-parties/${ partyId }`)
}

export async function requestVote(params: {
    locState: ClosedLoc | ClosedCollectionLoc,
    signer: Signer,
    callback: SignCallback,
}): Promise<string> {
    const { locState, signer, callback } = params;

    const client = locState.locsState().client;
    const api = client.nodeApi;
    const locData = locState.data();
    const locId = locData.id;
    const submittable = api.tx.vote.createVoteForAllLegalOfficers(locId.toDecimalString());
    const result = await signer.signAndSend({
        signerId: locState.data().ownerAddress,
        submittable,
        callback
    });
    const voteCreated = result.events.find(event => event.name === "VoteCreated" && event.section === "vote");
    if(!voteCreated) {
        throw new Error("Unable to retrieve vote ID");
    }
    const voteCreatedData = voteCreated.data as AnyJson[];
    return asString(voteCreatedData[0]);
}
