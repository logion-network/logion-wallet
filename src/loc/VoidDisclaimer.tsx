import { LocData } from "@logion/client";
import { LocType, UUID } from "@logion/node-api";
import DangerFrame from "src/common/DangerFrame";
import Icon from "src/common/Icon";
import InlineDateTime from "src/common/InlineDateTime";
import NewTabLink from "src/common/NewTabLink";

import "./VoidDisclaimer.css";

export interface Props {
    loc: LocData;
    detailsPath: (locId: UUID, type: LocType) => string;
}

export default function VoidDisclaimer(props: Props) {
    const { loc, detailsPath } = props;

    return (
        <>
            {
                loc.voidInfo !== undefined && loc.locType !== 'Collection' &&
                <DangerFrame
                    className="VoidDisclaimer loc-is-void"
                    title={ <span><Icon icon={ { id: 'void' } } width="45px" /> This LOC is VOID</span> }
                >
                    <p><strong>You have voided this LOC at the following date:</strong> <InlineDateTime
                        dateTime={ loc.voidInfo?.voidedOn } /></p>
                    <p><strong>Reason:</strong> { loc.voidInfo?.reason || "-" }</p>
                    {
                        loc.voidInfo.replacer !== undefined &&
                        <p><strong>This VOID LOC has been replaced by the following LOC: </strong>
                            <NewTabLink
                                href={ detailsPath(loc.voidInfo.replacer, loc.locType) }
                                iconId="loc-link"
                                inline
                            >
                                { loc.voidInfo.replacer.toDecimalString() }
                            </NewTabLink>
                        </p>
                    }
                    {
                        loc.voidInfo.replacer === undefined &&
                        <p>Please note that its public certificate shows a "VOID" mention to warn people that the
                            content of the LOC is not valid anymore.</p>
                    }
                    {
                        loc.voidInfo.replacer !== undefined &&
                        <p>Please note that its public certificate shows a "VOID" mention to warn people that the
                            content of the LOC is not valid anymore.
                            People will be automatically redirected to the replacing LOC when accessing to the void
                            LOC
                            URL and a mention of the fact that
                            the replacing LOC supersedes the void LOC will be visible on both certificates.
                        </p>
                    }
                </DangerFrame>
            }
            {
                loc.voidInfo !== undefined && loc.locType === 'Collection' &&
                <DangerFrame
                    className="VoidDisclaimer loc-is-void"
                    title={ <span><Icon icon={ { id: 'void' } } width="45px" /> This Collection LOC with all its related Collection Items are VOID</span> }
                >
                    <p><strong>You have voided this Collection LOC with all its related Collection Items at the
                        following date:</strong> <InlineDateTime dateTime={ loc.voidInfo?.voidedOn } /></p>
                    <p><strong>Reason:</strong> { loc.voidInfo?.reason || "-" }</p>
                    <p>Please note that related public certificates show a "VOID" mention to warn people that the
                        content of the Collection LOC as well as its related Collection Items are not valid
                        anymore.</p>
                </DangerFrame>
            }
        </>        
    );
}
