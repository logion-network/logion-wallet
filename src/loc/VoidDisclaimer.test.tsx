import { LocData } from "@logion/client";
import { UUID } from "@logion/node-api";

import { shallowRender } from "src/tests"
import VoidDisclaimer from "./VoidDisclaimer"

describe("VoidDisclaimer", () => {

    it("renders empty if not void", () => {
        const loc = {

        } as unknown as LocData;
        renderWithLoc(loc);
    })

    it("renders void info without replacer for transaction LOC", () => {
        const loc = {
            locType: "Transaction",
            voidInfo: voidInfoWithoutReplacer,
        } as unknown as LocData;
        renderWithLoc(loc);
    })

    it("renders void info with replacer for transaction LOC", () => {
        const loc = {
            locType: "Transaction",
            voidInfo: voidInfoWithReplacer,
        } as unknown as LocData;
        renderWithLoc(loc);
    })

    it("renders void info without replacer for identity LOC", () => {
        const loc = {
            locType: "Identity",
            voidInfo: voidInfoWithoutReplacer,
        } as unknown as LocData;
        renderWithLoc(loc);
    })

    it("renders void info with replacer for identity LOC", () => {
        const loc = {
            locType: "Identity",
            voidInfo: voidInfoWithReplacer,
        } as unknown as LocData;
        renderWithLoc(loc);
    })

    it("renders void info for collection LOC", () => {
        const loc = {
            locType: "Collection",
            voidInfo: voidInfoWithoutReplacer,
        } as unknown as LocData;
        renderWithLoc(loc);
    })
})

const replacer = new UUID("09abf4f9-a29b-4bce-b371-4360d38cad07");
const voidInfoWithReplacer = {
    voidedOn: "2022-09-28T17:25:00.000+02:00",
    reason: "Because",
    replacer,
}
const voidInfoWithoutReplacer = {
    voidedOn: "2022-09-28T17:25:00.000+02:00",
    reason: "Because",
}

function renderWithLoc(loc: LocData) {
    const element = shallowRender(<VoidDisclaimer loc={ loc } detailsPath={ () => "" } />);
    expect(element).toMatchSnapshot();
}
