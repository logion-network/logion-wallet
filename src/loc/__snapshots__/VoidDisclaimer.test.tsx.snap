// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoidDisclaimer renders empty if not void 1`] = `<React.Fragment />`;

exports[`VoidDisclaimer renders void info for collection LOC 1`] = `
<React.Fragment>
  <DangerFrame
    className="VoidDisclaimer loc-is-void"
    title={
      <span>
        <Icon
          icon={
            Object {
              "id": "void",
            }
          }
          width="45px"
        />
         This Collection LOC with all its related Collection Items are VOID
      </span>
    }
  >
    <p>
      <strong>
        You have voided this Collection LOC with all its related Collection Items at the following date:
      </strong>
       
      <InlineDateTime
        dateTime="2022-09-28T17:25:00.000+02:00"
      />
    </p>
    <p>
      <strong>
        Reason:
      </strong>
       
      Because
    </p>
    <p>
      Please note that related public certificates show a "VOID" mention to warn people that the content of the Collection LOC as well as its related Collection Items are not valid anymore.
    </p>
  </DangerFrame>
</React.Fragment>
`;

exports[`VoidDisclaimer renders void info with replacer for identity LOC 1`] = `
<React.Fragment>
  <DangerFrame
    className="VoidDisclaimer loc-is-void"
    title={
      <span>
        <Icon
          icon={
            Object {
              "id": "void",
            }
          }
          width="45px"
        />
         This LOC is VOID
      </span>
    }
  >
    <p>
      <strong>
        You have voided this LOC at the following date:
      </strong>
       
      <InlineDateTime
        dateTime="2022-09-28T17:25:00.000+02:00"
      />
    </p>
    <p>
      <strong>
        Reason:
      </strong>
       
      Because
    </p>
    <p>
      <strong>
        This VOID LOC has been replaced by the following LOC: 
      </strong>
      <NewTabLink
        href=""
        iconId="loc-link"
        inline={true}
      >
        12855903410953768523560335487932345607
      </NewTabLink>
    </p>
    <p>
      Please note that its public certificate shows a "VOID" mention to warn people that the content of the LOC is not valid anymore. People will be automatically redirected to the replacing LOC when accessing to the void LOC URL and a mention of the fact that the replacing LOC supersedes the void LOC will be visible on both certificates.
    </p>
  </DangerFrame>
</React.Fragment>
`;

exports[`VoidDisclaimer renders void info with replacer for transaction LOC 1`] = `
<React.Fragment>
  <DangerFrame
    className="VoidDisclaimer loc-is-void"
    title={
      <span>
        <Icon
          icon={
            Object {
              "id": "void",
            }
          }
          width="45px"
        />
         This LOC is VOID
      </span>
    }
  >
    <p>
      <strong>
        You have voided this LOC at the following date:
      </strong>
       
      <InlineDateTime
        dateTime="2022-09-28T17:25:00.000+02:00"
      />
    </p>
    <p>
      <strong>
        Reason:
      </strong>
       
      Because
    </p>
    <p>
      <strong>
        This VOID LOC has been replaced by the following LOC: 
      </strong>
      <NewTabLink
        href=""
        iconId="loc-link"
        inline={true}
      >
        12855903410953768523560335487932345607
      </NewTabLink>
    </p>
    <p>
      Please note that its public certificate shows a "VOID" mention to warn people that the content of the LOC is not valid anymore. People will be automatically redirected to the replacing LOC when accessing to the void LOC URL and a mention of the fact that the replacing LOC supersedes the void LOC will be visible on both certificates.
    </p>
  </DangerFrame>
</React.Fragment>
`;

exports[`VoidDisclaimer renders void info without replacer for identity LOC 1`] = `
<React.Fragment>
  <DangerFrame
    className="VoidDisclaimer loc-is-void"
    title={
      <span>
        <Icon
          icon={
            Object {
              "id": "void",
            }
          }
          width="45px"
        />
         This LOC is VOID
      </span>
    }
  >
    <p>
      <strong>
        You have voided this LOC at the following date:
      </strong>
       
      <InlineDateTime
        dateTime="2022-09-28T17:25:00.000+02:00"
      />
    </p>
    <p>
      <strong>
        Reason:
      </strong>
       
      Because
    </p>
    <p>
      Please note that its public certificate shows a "VOID" mention to warn people that the content of the LOC is not valid anymore.
    </p>
  </DangerFrame>
</React.Fragment>
`;

exports[`VoidDisclaimer renders void info without replacer for transaction LOC 1`] = `
<React.Fragment>
  <DangerFrame
    className="VoidDisclaimer loc-is-void"
    title={
      <span>
        <Icon
          icon={
            Object {
              "id": "void",
            }
          }
          width="45px"
        />
         This LOC is VOID
      </span>
    }
  >
    <p>
      <strong>
        You have voided this LOC at the following date:
      </strong>
       
      <InlineDateTime
        dateTime="2022-09-28T17:25:00.000+02:00"
      />
    </p>
    <p>
      <strong>
        Reason:
      </strong>
       
      Because
    </p>
    <p>
      Please note that its public certificate shows a "VOID" mention to warn people that the content of the LOC is not valid anymore.
    </p>
  </DangerFrame>
</React.Fragment>
`;
